### PR TITLE
Material can be opened for reading before the opening time

### PR DESCRIPTION
**Pull request by alalav1**
Modules have the `reading_opening_time` setting which allows the teacher to set time to open material for reading. The exercises in the module are not shown before the actual opening time.

### CODE REVIEW

**Comment by Plustest** [comment]
Can one of the admins verify this patch?

File: course/migrations/0046_coursemodule_reading_opening_time.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.2.3 on 2019-10-09 13:52
+
+from django.db import migrations, models
+import django.utils.timezone
+
+
+class Migration(migrations.Migration):
```
Comment (Mankro) [inline course/migrations/0046_coursemodule_reading_opening_time.py]: You don't need two migration files for adding one field.

**Comment by alalav1** [comment]
Javascript string translation is still missing (one string in chapter.js).

**Comment by Mankro** [comment]
When this is rebased against upstream master, we need to check the dependencies on the migrations. Migration files refer to to the previous migrations since they must be executed in the right order. The upstream master has new migrations and the migrations of this PR should come after those.

File: course/models.py [hunk 2]
```diff
@@ -685,6 +685,7 @@ class CourseModule(UrlMixin, models.Model):
     introduction = models.TextField(blank=True)
     course_instance = models.ForeignKey(CourseInstance, on_delete=models.CASCADE,
         related_name="course_modules")
+    reading_opening_time = models.DateTimeField(verbose_name=_("Opening time for reading"), null=True, blank=True)
```
Comment (Mankro) [inline course/models.py]: Is the verbose name of the field rendered in the form page? Or is there another keyword for the help string? The form page should display a help text that says that the reading opening time may be left empty if no material should be available before the module opening time.

File: course/models.py [hunk 3]
```diff
@@ -722,17 +723,28 @@ def clean(self):
...
             errors['late_submission_deadline'] = _("Late submission deadline must be later than the closing time.")
+        if self.reading_opening_time and self.reading_opening_time > self.opening_time:
+            errors['reading_opening_time'] = _("Opening time of reading material "
+                "must be earlier than the opening time of the exercises.")
         if errors:
             raise ValidationError(errors)
 
     def is_open(self, when=None):
         when = when or timezone.now()
+        if self.reading_opening_time:
```
Comment (Mankro) [inline course/models.py]: Are there any consequences that the module is considered to be open already during the reading time? Now, exercises may not be submitted even though this method says that the module is open.

File: course/permissions.py [hunk 4]
```diff
@@ -118,8 +118,6 @@ def has_permission(self, request, view):
         return True
 
     def has_object_permission(self, request, view, module):
-        if not isinstance(module, CourseModule):
```
Comment (Mankro) [inline course/permissions.py]: Why did you remove this `isinstance` check? Was there a problem with it?

File: course/permissions.py [hunk 5]
```diff
@@ -138,6 +136,11 @@ def has_object_permission(self, request, view, module):
             return module.are_requirements_passed(points)
         return True
 
+        if module.requirements.count() > 0:
```
Comment (Mankro) [inline course/permissions.py]: `if module.requirements.count() > 0:` is done two times in a row now. The second one should be removed, right?

File: course/templates/course/module.html [hunk 6]
```diff
@@ -43,7 +43,7 @@
 	</p>
 	{% endif %}
 
-  {% include "exercise/_children.html" with children=children accessible=True %}
+  {% include "exercise/_children.html" with children=children accessible=True exercise_accessible=current|is_open:now %}
```
[the remark this answers is not in the document]
Comment (Mankro) [reply course/templates/course/module.html]: Is `exercise_accessible` used for chapters too? The word exercise is often used for both exercises and chapters.

File: course/models.py [hunk 7]
```diff
@@ -722,17 +723,28 @@ def clean(self):
...
     def is_after_open(self, when=None):
         """
         Checks if current time is past the round opening time.
         """
+        when = when or timezone.now()
+        if self.reading_opening_time:
+            return self.reading_opening_time <= when
+        return self.opening_time <= when
+
+    def exercises_are_after_open(self, when=None):
```
Comment (Mankro) [inline course/models.py]: `exercises_are_after_open` is a bad name for the method. How about `have_exercises_been_opened`?

File: course/models.py (hunk 3 above)
Comment (Mankro) [reply course/models.py]: I see you added a separate method for checking the exercise opening time. I just wonder if old code uses the right method for each use case then.

File: course/templatetags/course.py [hunk 8]
```diff
@@ -81,6 +81,11 @@ def is_open(entry, now):
     return entry['opening_time'] <= now <= entry['closing_time']
 
 
+@register.filter
+def exercises_open(entry, now):
```
Comment (Mankro) [inline course/templatetags/course.py]: In this file, function `exercises_open` looks identical to `is_open`. Are both needed?

Comment (Mankro) [reply course/templatetags/course.py]: In the course module class, `is_open` uses the reading time, so now the `is_open` template tag is different when it uses the opening time. Inconsistency causes confusion.

File: exercise/templates/exercise/_children.html [hunk 9]
```diff
@@ -11,7 +11,17 @@
 {% endif %}
 {% elif entry|is_listed %}
 <li>
-  {% if accessible %}
+  {% if accessible and not entry.submittable%}
```
Comment (Mankro) [inline exercise/templates/exercise/_children.html]: Missing space after `entry.submittable`.

Comment (Mankro) [reply exercise/templates/exercise/_children.html]: Can you explain what these if statements do? I'd like to know what is used for chapters and exercises, and what happens before and after the opening time and reading opening time.

File: exercise/templatetags/exercise.py [hunk 10]
```diff
@@ -213,6 +213,18 @@ def min_group_size(context):
 
 @register.simple_tag(takes_context=True)
 def module_accessible(context, entry):
+    t = entry.get('reading_opening_time', entry.get('opening_time'))
```
Comment (Mankro) [inline exercise/templatetags/exercise.py]: Your new function `exercise_accessible` seems to be almost identical to the old `module_accessible`? What is the difference and is it really necessary to duplicate code? What if you add new parameters to the old function?

File: locale/fi/LC_MESSAGES/django.po [hunk 11]
```diff
@@ -268,32 +268,42 @@ msgstr "Huolto"
 msgid "Input an URL identifier for this module."
 msgstr "Syötä URL-osoitteessa käytettävä tunnus moduulille."
 
-#: course/models.py:699
+#: course/models.py:688
+msgid "Opening time for reading"
+msgstr "Lukumateriaalin avautuminen"
```
Comment (Mankro) [inline locale/fi/LC_MESSAGES/django.po]: "avautuminen" vai "avautumisaika" vai jotain muuta?

**Comment by Mankro** [comment]
This should be rebased against upstream master. The commit history must be cleaned. Database migrations and Finnish translations must be cleaned. Currently, changes in the translation django.po file cause git merge conflicts, but they should be easy to fix by runnig `python manage.py makemessages` again.

Note previous comments about migrations and the missing JavaScript string translation.

File: course/templatetags/course.py (hunk 8 above)
Comment (alalav1) [reply course/templatetags/course.py]: The method `is_open` has been removed.

File: exercise/templatetags/exercise.py (hunk 10 above)
Comment (alalav1) [reply exercise/templatetags/exercise.py]: The difference is whether the function uses the `reading_opening_time` or the `opening_time`. Those functions have now their common parts in the `is_accessible`.

File: course/models.py (hunk 2 above)
Comment (alalav1) [reply course/models.py]: Done.

File: course/permissions.py (hunk 5 above)
Comment (alalav1) [reply course/permissions.py]: Done.

File: course/templates/course/module.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (alalav1) [reply course/templates/course/module.html]: Exercise accessible uses the `opening_time` and is used to modify links and other elements.

File: exercise/templates/exercise/_children.html (hunk 9 above)
Comment (alalav1) [reply exercise/templates/exercise/_children.html]: `entry.submittable` checks if the children is exercise or chapter.

File: exercise/templatetags/exercise.py [hunk 12]
```diff
@@ -213,14 +213,14 @@ def min_group_size(context):
...
+def exercise_accessible(context, entry):
     t = entry.get('opening_time')
-    if t and t > _prepare_now(context):
-        return False
-    if entry.get('requirements'):
-        points = _prepare_context(context)
-        module = CourseModule.objects.get(id=entry['id'])
-        return module.are_requirements_passed(points)
-    return True
+    return is_accessible(context, entry, t)
```
Comment (Mankro) [inline exercise/templatetags/exercise.py]: It is confusing that `context` and `entry` are given to the `is_accessible` function call in a different order than in the definition of the function.

Comment (Mankro) [reply exercise/templatetags/exercise.py]: Looks like that it is OK to switch the order of the two variables in the definition of the `is_accessible` function. The `context` variable has been the first one in other functions too.

File: course/templates/course/_siblings.html [hunk 13]
```diff
@@ -4,7 +4,11 @@
 <nav class="row siblings">
 
     {% if next %}
-    {% module_accessible next as accessible %}
+     {% if next.submittable %}
```
Comment (Mankro) [inline course/templates/course/_siblings.html]: The indentation is wrong here. Don't use a single space. Our indentation in HTML templates has been inconsistent, but you could follow the style already used in the same file. Here, you could keep `{% if next.submittable %}` on the same level with `{% if next %}` and indent `{% exercise_accessible next as accessible %}` with four additional spaces so that it shows it is inside the if.

File: exercise/templates/exercise/_children.html [hunk 14]
```diff
@@ -11,18 +11,17 @@
 {% endif %}
 {% elif entry|is_listed %}
 <li>
-  {% if accessible %}
-  {% if entry.is_empty %}
-  {{ entry.name|parse_localization }}
+  {% if accessible and not entry.submittable or exercise_accessible %}
+   {% if entry.is_empty %}
```
Comment (Mankro) [inline exercise/templates/exercise/_children.html]: It looks like you have only changed one if condition here, but because you changed the indentations, many lines show up here in the git diff. Using one space is not good for indentation, and it is good to avoid making such whitespace changes that pollute the diff unless there is a good reason to do it. Either fix the indentations properly or do not change them on existing lines. I think you should remove your indentation changes so that the diff shows only the one changed if condition in the end after squashing.

File: exercise/static/exercise/chapter.js [hunk 15]
```diff
@@ -311,8 +311,18 @@
...
 			 } else {
 				$.ajax(this.url, {dataType: "html"})
-					.fail(function() {
-						exercise.showLoader("error");
+					.fail(function(jqXHR, textStatus, errorThrown) {
+						if (jqXHR.status == 403) {
+							exercise.hideLoader();
+							exercise.update($(
+								'<div id="exercise-all"><div class="info admonition" style="margin: 0px">' +
+								_('There will be an exercise when the module opens for submissions.') +
```
Comment (Mankro) [inline exercise/static/exercise/chapter.js]: `'There will be an exercise when the module opens for submissions.'` sounds a little off.
`'There will be an exercise HERE when the module opens for submissions.'` is slightly better. This could still be improved.

File: assets/js-translations/exercise.fi.json [hunk 16]
```diff
@@ -0,0 +1,4 @@
+{
+  "There will be an exercise when the module opens for submissions.":
+  "Tähän tulee tehtävä kierroksen avautuessa."
```
Comment (Mankro) [inline assets/js-translations/exercise.fi.json]: The Finnish string is OK, but the English string requires improvement.

File: course/templates/course/_siblings.html [hunk 17]
```diff
@@ -4,7 +4,11 @@
...
+      {% exercise_accessible next as accessible %}
```
Comment (Mankro) [inline course/templates/course/_siblings.html]: (I can't target this comment to the bottom of the file.) In the end of this file, there is the link to the "previous" item. This part is for the "next" item. Did you test the previous link? It could require the same separation of submittable and non-submittable items. Now, it always uses the `module_accessible` value.

File: exercise/templatetags/exercise.py [hunk 18]
```diff
@@ -278,6 +279,17 @@ def get_format_info(format):
     except KeyError as e:
         raise RuntimeError('Invalid format: \'{}\''.format(format)) from e
 
+
 @register.simple_tag
 def get_format_info_list(formats):
     return [get_format_info(format) for format in formats.split()]
+
+
+def is_accessible(entry, context, t):
```
Comment (Mankro) [inline exercise/templatetags/exercise.py]: Let's name the function `_is_accessible` and move it to the top of the file since the other helper functions are defined there. The underscore in the beginning implies that the function is not intended to be used outside this module (private function).

File: exercise/permissions.py [hunk 19]
```diff
@@ -31,6 +31,10 @@ def is_object_visible(self, request, view, exercise):
             self.error_msg(_("The exercise is not currently visible."))
             return False
 
+        if exercise.is_submittable and not exercise.course_module.have_exercises_been_opened():
+            self.error_msg(_("The exercise opens for submissions in {}.".format(exercise.course_module.opening_time)))
```
Comment (Mankro) [inline exercise/permissions.py]: I suspect that the date is formatted in the wrong way in the translated string. Shouldn't the `format` method be called on the string only after the `_` function has been called? Now you are trying to translate a string that contains a specific date value. Does it really work at all when you test it? The trivial fix is to move the parenthesis:
`_("The exercise opens for submissions in {}.").format(exercise.course_module.opening_time)`

**Comment by Mankro** [comment]
@alalav1 when you make these small changes, do not squash the commits yet into the original commit so that I can see the latest changes. Preferably, keep these new changes in one commit.

File: exercise/templatetags/exercise.py (hunk 12 above)
Comment (alalav1) [reply exercise/templatetags/exercise.py]: Switched parameters in the definition of `is_accessible` function.

File: course/templates/course/_siblings.html (hunk 13 above)
Comment (alalav1) [reply course/templates/course/_siblings.html]: Identation is changed so that the number of spaces in the beginning of the row is multiple of 4.

File: exercise/templates/exercise/_children.html (hunk 14 above)
Comment (alalav1) [reply exercise/templates/exercise/_children.html]: In August there was a problem when my atom changed spaces in tabs or other way around when I saved this file. I tried to fix them manually then. I'll use the proper indentation this time but git diff might still be little dirty.

File: course/templates/course/_siblings.html (hunk 17 above)
Comment (alalav1) [reply course/templates/course/_siblings.html]: Added the same separation between exercises and modules.

File: exercise/templatetags/exercise.py (hunk 18 above)
Comment (alalav1) [reply exercise/templatetags/exercise.py]: Renamed.

File: exercise/permissions.py (hunk 19 above)
Comment (alalav1) [reply exercise/permissions.py]: Moved the format out of the function.

File: exercise/static/exercise/chapter.js (hunk 15 above)
Comment (alalav1) [reply exercise/static/exercise/chapter.js]: "An exercise will be opened here once the round opens for submissions."?
Something else?
Or do you prefer your suggestion? @Mankro

File: assets/js-translations/exercise.fi.json (hunk 16 above)
Comment (alalav1) [reply assets/js-translations/exercise.fi.json]: New ideas:

Tässä julkaistaan tehtävä kierroksen avautuessa.

An exercise will be published here once the round opens for submissions.

Comment (Mankro) [reply assets/js-translations/exercise.fi.json]: I approve these.

File: exercise/static/exercise/chapter.js (hunk 15 above)
Comment (Mankro) [reply exercise/static/exercise/chapter.js]: See new ideas under assets/js-translations/exercise.fi.json
